### PR TITLE
Reduce payer balance needed to deploy programs

### PR DESCRIPTION
**Pull request by jstarry**
#### Problem
Program deploys require more SOL than necessary for a successful deploy.

The payer balance is used to pay for programdata account creation before the payer is credited for the deploy buffer account. The buffer account balance is about the same as the required programdata balance, so a payer basically has to cover twice the rent cost than they need to.

#### Summary of Changes
- Credit buffer balance before paying for programdata account creation

Fixes #

### CODE REVIEW

**Comment by jstarry** [comment]
Blocked on https://github.com/solana-labs/solana/issues/18629

**Comment by jackcmay** [comment]
I'm working on #18629

**Comment by jackcmay** [comment]
> I'm working on #18629

https://github.com/solana-labs/solana/pull/19750

**Comment by jackcmay** [comment]
Btw, if this was blocked I would expect a test that shows that :-)

File: programs/bpf_loader/src/lib.rs [hunk 1]
```diff
@@ -390,13 +390,27 @@ fn process_loader_upgradeable_instruction(
...
+
+            let mut instruction = system_instruction::create_account(
                 payer.unsigned_key(),
                 programdata.unsigned_key(),
                 1.max(rent.minimum_balance(programdata_len)),
                 programdata_len as u64,
                 program_id,
             );
+
+            // pass an extra account to avoid the overly strict UnbalancedInstruction error
```
Comment (jstarry) [inline programs/bpf_loader/src/lib.rs:409]: FYI @jackcmay I'm bumping into this issue now: https://github.com/solana-labs/solana/issues/9711

**Comment by jstarry** [comment]
This is ready for review @jackcmay, as mentioned above I hit the overly restrictive balanced instruction check but it's easy to work around for now

**Comment by jackcmay** [comment]
> This is ready for review @jackcmay, as mentioned above I hit the overly restrictive balanced instruction check but it's easy to work around for now

We are expecting to solve the overly restrictive balanced instruction issue in ABIv2 (FYI @Lichtso )

**Review by jackcmay — DISMISSED** [review]

**Pull request merged**